src/effects/backends/builtin/compressoreffect.cpp

```diff
@@ -0,0 +1,215 @@
+        const mixxx::EngineParameters& engineParameters)
+        : EffectState(engineParameters),
+          previousMakeUpGain(0),
+          previousStateDB(0) {
+}
+
+void CompressorEffect::loadEngineEffectParameters(
+        const QMap<QString, EngineEffectParameterPointer>& parameters) {
+    m_pThreshold = parameters.value("threshold");
+    m_pRatio = parameters.value("ratio");
+    m_pKnee = parameters.value("knee");
+    m_pAttack = parameters.value("attack");
+    m_pRelease = parameters.value("release");
+    m_pGain = parameters.value("gain");
+    m_pAutoMakeUp = parameters.value("automakeup");
+    m_pClipping = parameters.value("clipping");
+}
+
+void CompressorEffect::processChannel(
+        CompressorGroupState* pState,
+        const CSAMPLE* pInput,
+        CSAMPLE* pOutput,
+        const mixxx::EngineParameters& engineParameters,
+        const EffectEnableState enableState,
+        const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(enableState);
+
+    SINT numSamples = engineParameters.samplesPerBuffer();
+
+    // Compression
+    applyCompression(pState, engineParameters, pInput, pOutput);
+
+    // Auto make up
+    if (m_pAutoMakeUp->toInt() == AutoMakeUpOn) { 
+        applyAutoMakeUp(pState, pOutput, numSamples);
+    }
+
+    // Output gain
+    CSAMPLE gainParamDB = static_cast<CSAMPLE>(m_pGain->value());
+    SampleUtil::applyGain(pOutput, db2ratio(gainParamDB), numSamples);
+
+    // Clipping
+    if (m_pClipping->toInt() == ClippingOn) {
+        SampleUtil::copyClampBuffer(pOutput, pOutput, numSamples);
+    }
+}
+
+void CompressorEffect::applyAutoMakeUp(CompressorGroupState* pState, CSAMPLE* pOutput, const SINT& numSamples) {
+    CSAMPLE makeUpStateDB = pState->previousMakeUpGain;
+    CSAMPLE maxSample = SampleUtil::maxAbsAmplitude(pOutput, numSamples);
+    if (maxSample > 0) {
+        CSAMPLE maxSampleDB = ratio2db(maxSample);
+        CSAMPLE minGainReductionDB = -maxSampleDB + makeUpTarget;
+        makeUpStateDB = makeUpAttackCoeff * minGainReductionDB + (1 - makeUpAttackCoeff) * makeUpStateDB;
+        CSAMPLE levelDB = makeUpStateDB + maxSampleDB;
+        // logarithmic smoothing
+        if (levelDB > -1.0) {
+            makeUpStateDB = log10(levelDB + 2) - 1 - maxSampleDB;
+        }
+
+        pState->previousMakeUpGain = makeUpStateDB;
+        SampleUtil::applyGain(pOutput, db2ratio(makeUpStateDB), numSamples);
+    }
+}
+
+void CompressorEffect::applyCompression(CompressorGroupState* pState, const mixxx::EngineParameters& engineParameters, const CSAMPLE* pInput, CSAMPLE* pOutput) {
+    CSAMPLE thresholdParam = static_cast<CSAMPLE>(m_pThreshold->value());
+    CSAMPLE ratioParam = static_cast<CSAMPLE>(m_pRatio->value());
+    CSAMPLE kneeParam = static_cast<CSAMPLE>(m_pKnee->value());
+    CSAMPLE kneeHalf = kneeParam / 2.0f;
+    CSAMPLE attackCoeff = exp(-1000.0 / (m_pAttack->value() * engineParameters.sampleRate()));
```

Comment:
I am afraid exp() is slow can we store and reuse the result?

Reply:
alternatively, implement a good-enough approximation? see https://www.schraudolph.org/pubs/Schraudolph99.pdf

Reply:
Actually we can't store it, because user can change attack time in any time during playback.

I believe the compiler already has fast enough approximation, hasn't it? Because, for example, in Java world it isn't a problem, but I'm not sure about C++.

Reply:
The compiler use the math exact method by default. 
You may check if the user values have changed. The comparison is fast.

Reply:
Indeed, you're right. Finally I did caching of these coeffs and compare it to user values.